File: 0525. Contiguous Array.java
// Medium
// PrefixSum
// Time:O(n), Space:O(n)
// https://leetcode.cn/problems/contiguous-array/

import java.util.*;

class Solution {
    public int findMaxLength(int[] nums) {
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, -1);

        int maxLen = 0, sum = 0;

        for (int i = 0; i < nums.length; i++) {
            sum += (nums[i] == 1 ? 1 : -1); // if 0, add -1, so that sum for 1 and -1 would be 0

            if (map.containsKey(sum)) {
                maxLen = Math.max(maxLen, i - map.get(sum));
            } else {
                map.put(sum, i);
            }
        }
        return maxLen;
    }
}

File: 0525. Contiguous Array.py
from typing import List

class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        # {prefixsum, first appear index}
        prefix = {0: -1}

        prefix_sum = 0
        max_len = 0

        for i, num in enumerate(nums):
            # serve 0 as -1
            prefix_sum += 1 if num == 1 else -1

            if prefix_sum in prefix:
                len = i - prefix[prefix_sum]
                max_len = max(max_len, len)
            else:
                prefix[prefix_sum] = i;

        return max_len
    